arch/sim/src/sim/posix/sim_host_v4l2.c

```diff
@@ -0,0 +1,353 @@
+{
+  if (vdev != NULL)
+    {
+      close(vdev->fd);
+    }
+
+  free(vdev);
+  return 0;
+}
+
+int host_video_start_capture(struct host_video_dev_s *vdev)
+{
+  struct v4l2_buffer buf;
+  struct v4l2_requestbuffers reqbuf;
+  enum v4l2_buf_type type;
+  int i;
+
+  /* VIDIOC_REQBUFS initiate user pointer I/O */
+
+  reqbuf.type   = V4L2_BUF_TYPE_VIDEO_CAPTURE;
+  reqbuf.memory = V4L2_MEMORY_MMAP;
+  reqbuf.count  = MAX_REQBUFS;
+
+  if (-1 == host_video_ioctl(vdev->fd, VIDIOC_REQBUFS, &reqbuf))
+    {
+      perror("VIDIOC_REQBUFS");
+      return -errno;
+    }
+
+  if (reqbuf.count < 2)
+    {
+      errno = ENOMEM;
+      perror("Not enough buffers");
+      return -ENOMEM;
+    }
+
+  for (i = 0; i < reqbuf.count; i++)
+    {
+      memset(&buf, 0, sizeof(buf));
+      buf.type = V4L2_BUF_TYPE_VIDEO_CAPTURE;
+      buf.memory = V4L2_MEMORY_MMAP;
+      buf.index = i;
+      if (-1 == host_video_ioctl(vdev->fd, VIDIOC_QUERYBUF, &buf))
+        {
+          perror("VIDIOC_QUERYBUF");
+          return -errno;
+        }
+
+      vdev->addrs[i] = mmap(NULL, buf.length, PROT_READ | PROT_WRITE,
+                           MAP_SHARED, vdev->fd, buf.m.offset);
+      vdev->buflen[i] = buf.length;
+      if (vdev->addrs[i] == MAP_FAILED)
+        {
+          perror("Mmap failed");
+          goto err_out;
+        }
+
+      if (-1 == host_video_ioctl(vdev->fd, VIDIOC_QBUF, &buf))
+        {
+          perror("VIDIOC_QBUF");
+          goto err_out;
+        }
+    }
+
+  type = V4L2_BUF_TYPE_VIDEO_CAPTURE;
+  if (-1 == host_video_ioctl(vdev->fd, VIDIOC_STREAMON, &type))
+    {
+      perror("VIDIOC_STREAMON");
+      goto err_out;
+    }
+
+  return 0;
+err_out:
+  while (i--)
+    {
+      munmap(vdev->addrs[i], vdev->buflen[i]);
+    }
+
+  return -errno;
+}
+
+int host_video_stop_capture(struct host_video_dev_s *vdev)
+{
+  enum v4l2_buf_type type;
+  int i;
+
+  type = V4L2_BUF_TYPE_VIDEO_CAPTURE;
+  if (-1 == host_video_ioctl(vdev->fd, VIDIOC_STREAMOFF, &type))
+    {
+      perror("VIDIOC_STREAMOFF");
+      return -errno;
+    }
+
+  for (i = 0; i < MAX_REQBUFS; i++)
+    {
+      if (vdev->addrs[i] != NULL && vdev->addrs[i] != MAP_FAILED)
+        {
+          munmap(vdev->addrs[i], vdev->buflen[i]);
+        }
+    }
+
+  return 0;
+}
+
+int host_video_set_fmt(struct host_video_dev_s *vdev,
+                       uint16_t width, uint16_t height, uint32_t fmt,
+                       uint32_t denom, uint32_t numer)
+{
+  struct v4l2_format v4l2_fmt;
+  struct v4l2_streamparm streamparm;
+
+  memset(&v4l2_fmt, 0, sizeof(v4l2_fmt));
+  v4l2_fmt.type = V4L2_BUF_TYPE_VIDEO_CAPTURE;
+  v4l2_fmt.fmt.pix.width = width;
+  v4l2_fmt.fmt.pix.height = height;
+  v4l2_fmt.fmt.pix.pixelformat = fmt;
+  v4l2_fmt.fmt.pix.field = V4L2_FIELD_NONE;
+
+  if (-1 == host_video_ioctl(vdev->fd, VIDIOC_S_FMT, &v4l2_fmt))
+    {
+      perror("VIDIOC_S_FMT");
+      return -errno;
+    }
+
+  memset(&streamparm, 0, sizeof(streamparm));
```

Comment:
should we skip if denom == 0 like host_video_try_fmt

Reply:
denom == 0 will cause most v4l2 drivers to use default value instead, so it may be fine for set_fmt?

Reply:
Ok